source/client/process_impl.cc

```diff
@@ -128,6 +131,14 @@ void ProcessImpl::shutdown() {
 
   {
     auto guard = std::make_unique<Envoy::Thread::LockGuard>(workers_lock_);
+    // flush_worker_->shutdown() needs to happen before workers_.clear() so that
+    // metrics defined in workers scope will be included in the final stats
+    // flush which happens in FlushWorkerImpl::shutdownThread() after
+    // flush_worker_->shutdown() is called. For the order between worker shutdown() and
+    // shutdownThread(), see worker_impl.cc.
+    if (flush_worker_) {
```

Comment:
What guarantees that workers won't collect more statistics between the times when the flush worker was shut down and the time the workers get shut below?

Reply:
In `runInternal()`, we first call `waitForCompletion()` on all workers and then call `waitForCompletion()` on flush worker.
```
...
for (auto& w : workers_) {
    w->waitForCompletion();
  }

  if (!options_.statsSinks().empty() && flush_worker_ != nullptr) {
    // Stop the running dispatcher in flush_worker_. Need to be called after all
    // client workers are complete so that all the metrics can be flushed.
    flush_worker_->exitDispatcher();
    flush_worker_->waitForCompletion();
  }
...
```
On one hand, since `waitForCompletion()` only finished after all workers `work()` finished (where sending requests happens), so after workers' `waitForCompletion()` finished, there will be no requests generated.

On the other hand,  flush worker is shut down only happens after `runInternal()` finished which is after all workers' and flush worker's `waitForCompletion()` finished.